### PR TITLE
adding stop_for_best reduce(#26357)

### PR DESCRIPTION
**Pull request by MrPresent-Han**
related: #26357

### CODE REVIEW

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

File: internal/proxy/task_query.go [hunk 1]
```diff
@@ -127,15 +128,24 @@ func filterSystemFields(outputFieldIDs []UniqueID) []UniqueID {
 // parseQueryParams get limit and offset from queryParamsPair, both are optional.
 func parseQueryParams(queryParamsPair []*commonpb.KeyValuePair) (*queryParams, error) {
 	var (
-		limit  int64
-		offset int64
-		err    error
```
Comment (MrPresent-Han) [inline internal/proxy/task_query.go:132]: milvus has an underlying constraint that if there's no limit set, then no considering offset. But reduceStopForBest is independent from limit or offset

File: internal/proxy/task_query.go [hunk 2]
```diff
@@ -543,22 +542,29 @@ func reduceRetrieveResults(ctx context.Context, retrieveResults []*internalpb.Re
 	idSet := make(map[interface{}]struct{})
 	cursors := make([]int64, len(validRetrieveResults))
 
+	realLimit := typeutil.Unlimited
 	if queryParams != nil && queryParams.limit != typeutil.Unlimited {
-		loopEnd = int(queryParams.limit)
-
```
Comment (MrPresent-Han) [inline internal/proxy/task_query.go:548]: if we want reduce stop until best, then we should not set loopEnd to limit which may terminate the loop too early

File: internal/querynodev2/segments/result.go [hunk 3]
```diff
@@ -285,6 +286,7 @@ func MergeInternalRetrieveResult(ctx context.Context, retrieveResults []*interna
 			typeutil.AppendPKs(ret.Ids, pk)
```
Comment (MrPresent-Han) [inline internal/querynodev2/segments/result.go:286]: here, we only accumulate j when there is really one more result appended, if it's a duplicated pk, then not accumulate j

File: internal/proto/internal.proto [hunk 4]
```diff
@@ -142,7 +142,7 @@ message RetrieveRequest {
   int64 limit = 11; // Optional
   bool ignoreGrowing = 12;
   bool is_count = 13;
-  int64 iteration_extension_reduce_rate = 14;
```
Comment (yah01) [inline internal/proto/internal.proto:145]: this will break the compatibility, during rolling upgrading

**Review by yah01 — CHANGES_REQUESTED** [review]

File: internal/proto/internal.proto (hunk 4 above)
Comment (MrPresent-Han) [reply internal/proto/internal.proto:145]: true, I overlooked this point

File: internal/proxy/task_query.go [hunk 5]
```diff
@@ -543,22 +542,29 @@ func reduceRetrieveResults(ctx context.Context, retrieveResults []*internalpb.Re
...
 				cursors[sel]++
 			}
 		}
 	}
+	reduceStopForBest := false
+	if queryParams != nil {
+		reduceStopForBest = queryParams.reduceStopForBest
+	}
 
 	for j := 0; j < loopEnd; j++ {
```
Comment (XuanYang-cn) [inline internal/proxy/task_query.go:566]: The code should be the same as querynodev2/segments/results.go to remove the smaller ts row when two PKs are identical.

But it seems two places are not the same, breaking our designment of "get the latest when PKs are identical" 

Please help add them back at L578

**Review by XuanYang-cn — CHANGES_REQUESTED** [review]

File: internal/proxy/task_query.go [hunk 6]
```diff
@@ -127,15 +128,24 @@ func filterSystemFields(outputFieldIDs []UniqueID) []UniqueID {
...
+		offset            int64
+		reduceStopForBest bool
+		err               error
 	)
+	reduceStopForBestStr, err := funcutil.GetAttrByKeyFromRepeatedKV(ReduceStopForBestKey, queryParamsPair)
+	// if reduce_stop_for_best is provided
+	if err == nil {
+		reduceStopForBest, err = strconv.ParseBool(reduceStopForBestStr)
+		if err != nil {
+			return nil, fmt.Errorf("%s [%s] is invalid", ReduceStopForBestKey, reduceStopForBestStr)
```
Comment (congqixia) [inline internal/proxy/task_query.go:141]: Use `merr` here

Comment (MrPresent-Han) [reply internal/proxy/task_query.go]: changed

File: internal/proto/internal.proto (hunk 4 above)
Comment (MrPresent-Han) [reply internal/proto/internal.proto:145]: changed

**Pull request merged**